### DISCUSSION TITLE
enh: proposal to address scanned PDF handling in OpenWebUI RAG

### DISCUSSION THREAD

**Discussion by kukjun** [open-webui/open-webui#7344] [Ideas]
### Problem

I recently encountered an issue when uploading scanned PDF files, which resulted in the following error message:

![image](https://github.com/user-attachments/assets/5f94678c-5134-488c-b74b-a293c78d567b)

Upon investigating the issue, I discovered that PyPDF cannot process image-based PDFs, which lack an embedded text layer. This limitation is problematic for the OpenWebUI RAG (Retrieval-Augmented Generation) feature, as it prevents the use of scanned PDFs.

To address this, I propose integrating OCR (Optical Character Recognition) libraries such as EasyOCR or Tesseract. The solution would involve the following approach:

First, PyPDF attempts to process the PDF as usual.
If PyPDF fails to extract text (indicating an image-based PDF), the system would then use an OCR library to extract text from the embedded images.

This fallback mechanism would ensure that scanned PDFs can also be utilized within OpenWebUI, significantly enhancing its functionality.

Before implementing and submitting a PR, I would like to confirm whether this approach aligns with the direction of OpenWebUI's development and if there are alternative suggestions for handling scanned PDFs.

I look forward to your feedback and insights! ☺️

If you agree with this approach, I will proceed to implement the solution and submit a PR accordingly.

**Comment by auliaihendradi** [comment]
It's a good fallback mechanism, I think it'd be better to incorporate it into setting also, to make sure it won't interrupt the existing flow or make sure people have 2 options/flows

**Reply by kukjun** [reply]
Thank you for suggesting such a good approach. I will implement an option in the user interface to let users choose whether to apply OCR.

**Comment by hgd000** [comment]
It is such a great idea. Since I am facing the same issue as you, the fallback mechanism would be an urgent need for me. Looking forward to see what the result is!!!

**Comment by Awee3** [comment]
can we use the "OCR Scanned Pdf" tools?

**Comment by smonegv** [comment]
Good idea, we are also working on this problem. We could share thoughts and the best way forward.

**Reply by smonegv** [reply]
For the moment we are considering to build a tool ad HOC without altering ( in this first phase ) the source code of openwebui

**Comment by rgaricano** [comment]
about pypdf, the problem is detected, it has problems with reshape images when OCR, there are a [PR](https://github.com/open-webui/open-webui/pull/11240)  to use other lib waiting for some more test ...
(of course that could be perfect have some more options than tika for use external tools, but it also can by implemented through functions or tools)

**Reply by smonegv** [reply]
Thank you, I also stand by for updates. In the meantime I continue with local testing and improvements. If you think of any updates or insights to share let us know. Also to do further testing.
